### ISSUE TITLE
Panic when trying to open file

### ISSUE THREAD

**Issue by qarmin** [pop-os/cosmic-edit#4]
Opening - [ABC.zip](https://github.com/pop-os/cosmic-text-editor/files/11082343/ABC.zip)

will crash with message

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `1`', /home/rafal/.cargo/git/checkouts/cosmic-text-92dc585b4973ff15/e788c17/src/shape.rs:447:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by KurtB03** [comment]
May i ask what kind of file this is even supposed to be ?
It's non-sensical when rendered with ASCII or UTF-8, and i can't recognize a BOM (Byte Order Mark) or Filetype signature.
Or which fonts are required to render it ?

**Comment by qarmin** [comment]
This file contains random utf-8 characters(should all be valid) that I used to test different app

**Comment by KurtB03** [comment]
```rust
    /// Will panic if `line` contains more than one paragraph.
    pub fn new(font_system: &mut FontSystem, line: &str, attrs_list: &AttrsList) -> Self {
        let mut spans = Vec::new();

        let bidi = unicode_bidi::BidiInfo::new(line, None);
        let rtl = if bidi.paragraphs.is_empty() {
            false
        } else {
            assert_eq!(bidi.paragraphs.len(), 1);
            let para_info = &bidi.paragraphs[0];
            let line_rtl = para_info.level.is_rtl();

            log::trace!("Line {}: '{}'", if line_rtl { "RTL" } else { "LTR" }, line);
```

If looked up the location of the crash in https://github.com/pop-os/cosmic-text It looks like there are issues with the coexistence of writing systems in different directions.
It would surely be nice to catch that crash, however mundane examples like `println!( "ויקימדיה");` work, so it's not a big problem especially considering the current state, without even clipboard support.

Still good to know.

PS.: i have no clue what the hebrew means, don't come for me, i'm just testing

**Comment by Dygear** [comment]
The problem is actually in the [cosmic-text](https://github.com/pop-os/cosmic-text) repo. That line is `assert_eq!(bidi.paragraphs.len(), 1);` and has since moved to [Line 694](https://github.com/pop-os/cosmic-text/blob/main/src/shape.rs#L694C13-L694C50) in the latest commit. Probably should be reported there and closed here, if not already fixed by now.

**Comment by Karlosbubi** [comment]
I've tried to dig trough it when it was new, it's not so clear if this issue is caused by cosmic-text or caused here and found by cosmic-text.
As the assert is a check and not a fault it self. I would be interesting how other consumers of cosmic-text handle this file.

**Comment by Dygear** [comment]
Yeah, I don't have a good answer for that. It's documented in the code.
> Will panic if `line` contains more than one paragraph.

That's exactly the assert it hits, and fails at. How does a line contain more than one paragraph ... I have no idea. An `\r` and or an `\n` should make a new line that would also make a new paragraph. But I'm probably getting hung up on the naming convention here. Specifically bidi with a quick Google search is to do with the bidirectional nature of text. I just don't know enough about the problem area to be useful. But it's not really a problem with cosmic editor so much as it's a known edge case of cosmic text.
